books/nlp/src/llms/rag/README.md

```diff
@@ -1 +1,169 @@
+<!-- markdownlint-disable-file MD033 -->
+
 # RAG
+
+{{ #aipr_header colab=nlp/rag.ipynb }}
+
+## Intro and Motivation for RAG
+
+After an LLM has been pre-trained, its learning is captured in _parametric_ knowledge.
+This speak is jargon simply implying that the knowledge is captured in the LLM's
+weight parameters. If the LLM is further fine-tuned for improved instruction following
+or alignment, these knowledge specializations are still also considered to be parametric
+in nature (i.e., since these involve weight parameter updates).
+
+However, researchers have observed that relying only on the LLM's parametric knowledge,
+can be suboptimal and this is especially observed when performing knowledge-intensive
+tasks. Some pundits have argued that long-tail knowledge is not easily captured
+in parametric form.
+
+To remedy this drawback of an LLM's parametric knowledge, we can consider providing
+an LLM with non-parametric knowledge. Retrieval-Augmented Generation (RAG) is one
+such technique that aims to provide knowledge in the form of additional context
+to an LLM at inference time. As it's name suggests, this method involves
+retrieving facts (i.e., knowledge) from a data store and augmenting (e.g., by
+string concatenation) the original prompt or query to the LLM with these facts.
+
+## Components of a RAG System
+
+A RAG system is comprised of three main components, namely:
+
+- **Knowledge Store** — contains non-parametric knowledge facts that the system
+  can use at inference time in order to produce more accurate responses to queries.
+- **Retriever** — a model that takes in a user query and retrieves the most relevant
+  knowledge facts from the knowledge store. (NOTE: the retriever is also used to
+  populate or index the knowledge store during setup.)
+- **Generator** — a model that takes in the user's query and additional context
+  and provides a response to that query.
+
+## Canonical RAG Pipeline
+
+The canonical pipeline for RAG is as follows:
+
+1. User submits a query to the RAG system
+2. **[Retrieval Step]** The RAG system matches the query with the relevant facts
+   from the knowledge store. The top k matched facts are retrieved.
+3. **[Generation Step]** The content of the retrieved facts are used to augment the
+   query and subsequently pass to the generator.
+4. Response is returned back to the user. (Post-processing steps may be applied
```

Comment:
- "Post-processing steps may be applied to the raw result from the generator prior to returning it to the user."